barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_format.cpp

```diff
@@ -230,45 +230,32 @@ void build_constraints(Builder& builder,
-            // verifier circuit requires that the indices to a nested proof aggregation state are a circuit constant.
-            // The user tells us they how they want these constants set by keeping the nested aggregation object
-            // attached to the proof as public inputs. As this is the only object that can prepended to the proof if the
-            // proof is above the expected size (with public inputs stripped)
+            // A proof passed into the constraint should be stripped of its inner public inputs, but not the nested
+            // aggregation object itself. The verifier circuit requires that the indices to a nested proof aggregation
+            // state are a circuit constant. The user tells us they how they want these constants set by keeping the
+            // nested aggregation object attached to the proof as public inputs. As this is the only object that can
+            // prepended to the proof if the proof is above the expected size (with public inputs stripped)
             std::array<uint32_t, HonkRecursionConstraint::AGGREGATION_OBJECT_SIZE> nested_aggregation_object = {};
-            // If the proof has public inputs attached to it, we should handle setting the nested aggregation object
-            // The public inputs attached to a proof should match the aggregation object in size
-            if (constraint.proof.size() - proof_size_no_pub_inputs !=
-                HonkRecursionConstraint::AGGREGATION_OBJECT_SIZE) {
-                auto error_string = format(
-                    "Public inputs are always stripped from proofs unless we have a recursive proof.\n"
-                    "Thus, public inputs attached to a proof must match the recursive aggregation object in size "
-                    "which is ",
-                    HonkRecursionConstraint::AGGREGATION_OBJECT_SIZE);
-                throw_or_abort(error_string);
-            }
+            const size_t inner_public_input_offset = 3;
+            info("constraint proof: ", constraint.proof);
             for (size_t i = 0; i < HonkRecursionConstraint::AGGREGATION_OBJECT_SIZE; ++i) {
                 // Set the nested aggregation object indices to the current size of the public inputs
                 // This way we know that the nested aggregation object indices will always be the last
                 // indices of the public inputs
-                nested_aggregation_object[i] = static_cast<uint32_t>(constraint.public_inputs.size());
+                nested_aggregation_object[i] = static_cast<uint32_t>(constraint.proof[inner_public_input_offset + i]);
                 // Attach the nested aggregation object to the end of the public inputs to fill in
                 // the slot where the nested aggregation object index will point into
-                constraint.public_inputs.emplace_back(constraint.proof[i]);
+                constraint.public_inputs.emplace_back(constraint.proof[inner_public_input_offset + i]);
```

Comment:
```suggestion
                constraint.public_inputs.emplace_back(nested_aggregation_object[i]);
```
Maybe we do this instead so that we know we are matched up with the `nested_aggregation_object` now that the nested agg object is no longer public input indices.